### ISSUE TITLE
Physics mul transform might be wrong?

### ISSUE THREAD

**Issue by Andrewwango** [deepinv/deepinv#489]
Currently, the DACLoss #410 uses maybe not the best implementation for a transformed physics:

```python
        physics2 = deepcopy(physics)
        A, A_adjoint, A_dagger = physics2.A, physics2.A_adjoint, physics2.A_dagger
        physics2.A = lambda x, *args, **kwargs: A(
            self.T_e.inverse(x, **e_params), *args, **kwargs
        )
        physics2.A_adjoint = lambda y, *args, **kwargs: self.T_e(
            A_adjoint(y, *args, **kwargs), **e_params
        )
        physics2.A_dagger = lambda y, *args, **kwargs: self.T_e(
            A_dagger(y, *args, **kwargs), **e_params
        )
```

It would be nice to have the much nicer implementation using `physics.__mul__`:

```python
        physics2 = physics * LinearPhysics(
            A=lambda x: self.T_e.inverse(x, **e_params),
            A_adjoint=lambda x: self.T_e(x, **e_params),
        )
```
However, when training with this using the setup in https://github.com/Andrewwango/ssibench/blob/9cd076dfe2de27f199c69306b5e341b38e733904/train.py#L117, the results are very different and much worse.

Something seems to be wrong with `LinearPhysics.__mul__`, but I don't know what. Maybe something from the original physics isn't moving over correctly (which does move over when doing deepcopy).

**Comment by jscanvic** [comment]
I don't have a definitive answer as to what is going on but I managed to make a minimal reproducible example of the bug. Hopefully it can help!

The issue seems to be in the implementation of `Physics.A_dagger` for the physics obtained using `LinearPhysics.__mul__`. From what I witnessed, A and A_adjoint behave the same for both physics.

```python
import torch
import deepinv as dinv
from copy import deepcopy

T_kwargs = {"x_shift": torch.tensor([1]), "y_shift": torch.tensor([1])}

def impl1(physics, T):
    physics_deepcopy = deepcopy(physics)
    A, A_adjoint, A_dagger = physics_deepcopy.A, physics_deepcopy.A_adjoint, physics_deepcopy.A_dagger
    physics_deepcopy.A         = lambda x, *args, **kwargs: A(
            T.inverse(x, **T_kwargs), *args, **kwargs
    )
    physics_deepcopy.A_adjoint = lambda y, *args, **kwargs:   T(
            A_adjoint(y, *args, **kwargs), **T_kwargs
    )
    physics_deepcopy.A_dagger  = lambda y, *args, **kwargs:   T(
            A_dagger(y, *args, **kwargs), **T_kwargs
    )
    return physics_deepcopy


def impl2(physics, T):
    physics_mul = physics * dinv.physics.LinearPhysics(
        A=lambda x: T.inverse(x, **T_kwargs),
        A_adjoint=lambda y: T(y, **T_kwargs),
    )
    return physics_mul

torch.manual_seed(0)

physics = dinv.physics.Tomography(angles=500, img_width=128)
T = dinv.transform.Shift()

physics_deepcopy = impl1(physics, T)
physics_mul = impl2(physics, T)

x = torch.randn(1, 1, physics.img_width, physics.img_width, dtype=torch.float32)
y = physics_deepcopy.A(x)
x_deepcopy = physics_deepcopy.A_dagger(y)
x_mul = physics_mul.A_dagger(y)

# L2 error: 64.98800659179688
print(f"L2 error: {torch.linalg.norm(x_deepcopy - x_mul)}")

assert torch.allclose(x_deepcopy, x_mul, rtol=1e-3, atol=1e-4)
```

**Comment by mh-nguyen712** [comment]
I just added a ComposedPhysics and ComposeLinearPhysics in my recent PR #462 . Can you guys take a look? I thinks we should avoid using `lambda` function for our physics classes to keep them compatible with torch.nn.Module, as many physics.A has dependent params (which are tensors)

**Comment by Andrewwango** [comment]
Thanks @mh-nguyen712 i will take a look!

**Comment by mh-nguyen712** [comment]
I think we should avoid using `lambda` for changing the function attribute of Physics, since there might be some dependent parameters which make the class inheritance fails (e.g., if we're doing this on DecomposablePhysics, other methods could fail since we dont update the `mask` , ...).
For example, following the code snippet of @jscanvic, the following check fails:
```
physics = dinv.physics.Inpainting(tensor_size=(3,64,64), mask = 0.5)

physics_deepcopy = impl1(physics, T)

torch.allclose(physics_deepcopy.A_adjoint_A(x), physics_deepcopy.A_adjoint(physics_deepcopy.A(x)))

```

Moreover, regarding the pseudo-inverse, I think that in general we dont have `(A T^{-1})^+ = T A^+`, unless `Ker(TA) = Ker(A)`, am I right?

**Comment by jscanvic** [comment]
Erratum: A also needs to have full column rank apparently! This isn't the case for some of the physics like inpainting.

I'm not sure what we should make of it.

**Comment by mh-nguyen712** [comment]
The current ComposedPhysics we just update the A and A_adjoint and leave the A_dagger solved by conjugate_gradient. This is the general behavior if we do not have better way.

**Comment by tachella** [comment]
I agree with Hai, I think the bug is more in the DACLoss, which assumes that (AT)^dagger=T^{-1}A^dagger than in the multiplication of operators. There is no other safe solution than relying on conjugate gradient once we combine two linear operators.

**Comment by Andrewwango** [comment]
@mh-nguyen712 @tachella 

Can we add a `ComposedDecomposablePhysics`? My use case is, my original physics is decomposable, and my model (`MoDL`) requires the physics prox, which is available in closed for for decomposable physics. But when I transform it, I still want the closed form prox rather than the approximation provided by `LinearPhysics`. 

I think this is where my bug is coming from, as this bug wouldn't appear if you did the `physics_new = deepcopy(physics)` approach from my original bug, or you use a model that doesn't require the prox/dagger.

Otherwise, I will leave the DAC loss as is and close this issue (@tachella, could this resurface elsewhere i.e. not just in DAC loss e.g. when symmetrizing physics elsewhere in deepinv using the `__mul__` approach?)

**Comment by tachella** [comment]
@Andrewwango I think it will become overcomplex to add a `ComposedDecomposablePhysics` class for this specific use-case, which doesn't have any strong mathematical meaning (ie having an explicit SVD decomposition as in `DecomposablePhysics`. Nonetheless, I could imagine some cases where it is possible to easily obtain an SVD decomposition when composing with a transform (eg `Inpainting `+ circular `Shift`), but these should be handled on a per-case basis (as far as I know there is no automatic way of doing this).

Even in the case of composing with transformations, the resulting `A_dagger` and `prox` can be non-trivial if there is zero-padding or interpolation going on in the transform. I think the safe choice is to always fall back to conjugate gradient solvers in these cases.

I'm also not sure the `Modl` example - as long as you use multiple coils, the closed-form expressions should be gone?

**Comment by Andrewwango** [comment]
Ok, maybe what I want is not appropriate for a "Composed" physics, but rather a physics where you can hijack A and A_adjoint without modifying anything else about the physics. I will keep DAC loss as is, as long as warnings and caveats are put in #462 

In the MoDL example, if I have `physics = MRI` (i.e. decomposable physics), then I want to preserve my closed-form expressions after transforming. Of course if I have `physics = MultiCoilMRI` then this isn't an issue as it's linear physics already.

@tachella If we don't fix this, you will run into the same error in your symmetrized reconstructor, if `physics` is decomposable it will lose its closed form expressions:

https://github.com/deepinv/deepinv/blob/4da2c0b0286219d484892933fce9d1bcfd7fad43/deepinv/transform/base.py#L307

**Comment by tachella** [comment]
I agree with hijacking specific cases, as long as we add some warnings or we make sure that we limit the usage to certain operators :)

**Comment by Andrewwango** [comment]
Adding wontfix label for now, let's re-discuss when symmetrised reconstructors is ready for review! And let's leave DACLoss as is for now.

**Comment by Andrewwango** [comment]
Closing in favour of continuing discussion in #533
